src/backend/cgcod.c

```diff
@@ -726,7 +726,7 @@ code *prolog()
     assert(usedalloca != 1);
     AllocaOff = alignsection(usedalloca ? (Foff - REGSIZE) : Foff, REGSIZE, bias);
 
-    CSoff = alignsection(AllocaOff - cstop * REGSIZE, REGSIZE, bias);
+    CSoff = alignsection(AllocaOff - (int)cstop * REGSIZE, REGSIZE, bias);
```

Comment:
> But it doesn't explain why the bug doesn't appear in dmd compiled by g++.

As explained above, there is no bug to be noticed easily. A few optimizations are disabled.
I just verified that g++ generates code and results as the debug version. The sign-extension issue doesn't happen anymore with a 64-bit build, though, because `cstop` is now `size_t`.